src/coreclr/jit/compiler.cpp

```diff
@@ -3325,6 +3325,30 @@ void Compiler::compInitOptions(JitFlags* jitFlags)
         opts.compJitSaveFpLrWithCalleeSavedRegisters = JitConfig.JitSaveFpLrWithCalleeSavedRegisters();
     }
 #endif // defined(DEBUG) && defined(TARGET_ARM64)
+
+#if defined(TARGET_AMD64)
+    rbmAllFloat            = RBM_ALLFLOAT_INIT;
+    rbmFltCalleeTrash      = RBM_FLT_CALLEE_TRASH_INIT;
+    rbmCntCalleeTrashFloat = CNT_CALLEE_TRASH_FLOAT_INIT;
+
+    if (DoJitStressEvexEncoding())
+    {
+        rbmAllFloat |= RBM_HIGHFLOAT;
+        rbmFltCalleeTrash |= RBM_HIGHFLOAT;
+        rbmCntCalleeTrashFloat += 16;
+    }
+    rbmCalleeTrash                    = RBM_CALLEE_TRASH_INIT;
+    rbmCalleeTrashNoGC                = RBM_CALLEE_TRASH_NOGC_INIT;
+    rbmCalleeTrashWriteBarrier        = RBM_CALLEE_TRASH_WRITEBARRIER_INIT;
+    rbmCalleeGCTrashWriteBarrier      = RBM_CALLEE_GCTRASH_WRITEBARRIER_INIT;
+    rbmCalleeTrashWriteBarrierByref   = RBM_CALLEE_TRASH_WRITEBARRIER_BYREF_INIT;
+    rbmCalleeGCTrashWriteBarrierByref = RBM_CALLEE_GCTRASH_WRITEBARRIER_BYREF_INIT;
+    rbmStopForGCTrash                 = RBM_STOP_FOR_GC_TRASH_INIT;
+    rbmInitPInvokeFrameTrash          = RBM_INIT_PINVOKE_FRAME_TRASH_INIT;
+    rbmProfilerEnterTrash             = RBM_PROFILER_ENTER_TRASH_INIT;
+    rbmProfilerLeaveTrash             = RBM_PROFILER_LEAVE_TRASH_INIT;
+    rbmProfilerTailcallTrash          = RBM_PROFILER_TAILCALL_TRASH_INIT;
+#endif // TARGET_AMD64
```

Comment:
> So these are global variables now, that are initialized every time the JIT is initialized?
> 
> Can we assume the memory write is atomic? Should they be initialized using InterlockedExchange64? Using Interlocked, we could assert they either have the same value as written, or zero (uninitialized).
> 
> I presume they are not on Compiler because either (1) it was inconvenient to plumb a Compiler* everywhere they are needed, (2) throughput, or (3) you couldn't/didn't want to use the TLS Compiler* (maybe only available in DEBUG?)

Yes, this is why --- passing `compiler` around was both clunky and also incurred its own overhead because it needed to be checked on each access. 

> 
> Could there be a scenario where these change for a single clrjit.dll instance? E.g., what if crossgen2/AOT compilation loads the JIT and compiles both an AVX-512 targeting function and a non-AVX-512 targeting function?

I could use some help here, because honestly I'm not familiar enough with the internals to understand this. As it's done now, I am assuming that these are unchanged and compiling for a single arch/ISA.

> 
> Currently, they are only changed with the JitStress mode. They will presumably be set based on ISA. Does this initialization location have all the appropriate ISA information?

The compiler object has this information as far as I know. Essentially, I would envision replacing with an `compOpportunisticallyDependsOn(InstructionSet_AVX512F_VL)` check.

@tannergooding @kunalspathak are my assumptions here correct? Will this be an issue with `crossgen`?

Reply:
I think it needs to be filtered on`compOpportunisticallyDependsOn(InstructionSet_AVX512F)`, but I agree that I would expect crossgen to be filtered correctly presuming it.

I do think this needs to be "per compiler instance" and not part of the global state as we can have AltJits or other scenarios where one `compiler` supports an ISA but another does not.

If we could cheapen this further, such as pulling the relevant state into a local where the register allocator needs it, that would presumably be "better".

If the TP regression is "too much" even with that, I could imagine a few ways we could use a template or inlining to ensure the LSRA code still uses a constant and is primarily filtered from a centralized high level check (of course, this comes at the cost of increased deployment size for clrjit.dll).

Reply:
> If we could cheapen this further, such as pulling the relevant state into a local where the register allocator needs it

The problem is it is not just LSRA that depends on them, but optimizer and codegen as well. Hence it is little hard to extract that out just at the LSRA level.

Reply:
> I presume they are not on Compiler because either (1) it was inconvenient to plumb a Compiler* everywhere they are needed, (2) throughput, or (3) you couldn't/didn't want to use the TLS Compiler* (maybe only available in DEBUG?)

The reason was `#1` , but what you have brought up is a legitimate concern. The only way to avoid it to have them on compiler object and then pass `compiler` instance around. Depending on which area are using these macros, we could do something like, where you have a macro (per file since they are not that many) which takes compiler object, but at all the places where it is needed, you use a version of macro that actually call the one that takes compiler object. That way, at the callsite, you don't have to pass `compiler` object. Its ugly :(

```
#define ABC_WITH_COMPILER(compiler) compiler->isEvex ? x : y
#define ABC ABC_WITH_COMPILER(this)

 = ABC;

```

Reply:
If I remember correctly, the throughput overhead of using a compiler instance variable was about 0.50% overhead. It was some of the ugliness of isolating that throughput to X64 that motivated the above approach, i.e. there were _many_ spots where we would have...

```C++
#if defined(TARGET_AMD64)
... = compiler->rbmCalleFltTrash
#else
... = RBM_CALLE_FLT_TRASH
#endif
```

I believe we can use the above approach to mitigate some of this but it will still be a little ugly...

```C++
#if defined(TARGET_AMD64)
#define RBM_CALLE_FLT_TRASH_USE compiler->rbmCalleFltTrash
#else
#define RBM_CALLE_FLT_TRASH_USE RBM_CALLE_FLT_TRASH
#endif
```

Reply:
@kunalspathak @tannergooding do you have any more thoughts on how to adjust the PR for this?

Reply:
Did you try the approach in https://github.com/dotnet/runtime/pull/79544#discussion_r1063850725 and https://github.com/dotnet/runtime/pull/79544#discussion_r1063883720?

Reply:
Ok, let me try this approach https://github.com/dotnet/runtime/pull/79544#discussion_r1063883720 then and see how the numbers look.

Reply:
@kunalspathak I have a draft of this up at https://github.com/dotnet/runtime/pull/80559/commits/9c386bc654bedeced1c05b66e9a0bbb4a581b70e . Currently waiting for test data / throughput.

To be honest, I do not like it, it seems really messy (even if we move the macro definitions to the call sites). It also seems the alternatives aren't great either.

Reply:
FYI, with the change https://github.com/dotnet/runtime/commit/9c386bc654bedeced1c05b66e9a0bbb4a581b70e ...

![image](https://user-images.githubusercontent.com/91913526/212180309-6965b96b-d6b2-421a-a913-719c8a31a742.png)

It seems like most approaches have been pretty similar with throughput impact.